GFortran 10 Argument Mismatch

@michaelkuhn reported an issue in Spack, which indicated that ADIOS2 needs the Fortran compiler flag `-fallow-argument-mismatch` flag in order to build with GCC/GFortran 10.

Note:
```
Using this option is strongly discouraged. It is possible to provide standard-conforming code which allows different types of arguments by using an explicit interface and TYPE(*). 
```
https://gcc.gnu.org/onlinedocs/gfortran/Fortran-Dialect-Options.html

**Describe the bug**
Please see https://github.com/spack/spack/pull/16570

**To Reproduce**
s. above

**Desktop (please complete the following information):**
 - OS/Platform: Ubuntu
 - Build: GCC 10
 - ADIOS version: 2.5.0

**Additional context**
See: https://github.com/spack/spack/pull/16570

Full error if anyone wants to fix (there might be errors in other files too.:
```bash
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:225:43:

  225 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:210:43:

  210 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(2)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:195:43:

  195 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(1)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:180:43:

  180 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(8)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:165:43:

  165 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:145:43:

  145 |             call adios2_attribute_data_f2c(dataMax, attribute%length, &
      |                                           1
......
  240 |             call adios2_attribute_data_f2c(data, length, attribute%f2c, ierr)
      |                                           2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(4096)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:114:44:

  114 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:101:44:

  101 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(2)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:88:44:

   88 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(1)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:75:44:

   75 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(8)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:62:44:

   62 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/INTEGER(8)).
/Users/usr/Downloads/adios2/source/bindings/Fortran/modules/adios2_attribute_data_mod.f90:48:44:

   48 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            1
......
  127 |             call adios2_attribute_value_f2c(data, attribute%f2c, ierr)
      |                                            2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (CHARACTER(0)/INTEGER(8)).
make[2]: *** [bindings/Fortran/CMakeFiles/adios2_fortran.dir/modules/adios2_attribute_data_mod.f90.o] Error 1
make[1]: *** [bindings/Fortran/CMakeFiles/adios2_fortran.dir/all] Error 2
make: *** [all] Error 2
```

--- a/bindings/Fortran/CMakeLists.txt
+++ b/bindings/Fortran/CMakeLists.txt
@@ -11,6 +11,10 @@ FortranCInterface_VERIFY(CXX QUIET)
 # Check whether the compiler supports Fortran submodule constructs we need.
 adios2_check_fortran_submodules(ADIOS2_HAVE_FORTRAN_SUBMODULES)
 
+if(ADIOS2_USE_Fortran_flag_argument_mismatch)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fallow-argument-mismatch")
+endif()
+
 add_library(adios2_fortran
   f2c/adios2_f2c_adios.cpp
   f2c/adios2_f2c_attribute.cpp

--- a/cmake/DetectOptions.cmake
+++ b/cmake/DetectOptions.cmake
@@ -352,4 +352,7 @@ if(ADIOS2_USE_Fortran)
   check_float_type_representation("real(kind=4)" REAL4_TYPE_Fortran LANGUAGE Fortran)
   check_float_type_representation("real(kind=8)" REAL8_TYPE_Fortran LANGUAGE Fortran)
   check_float_type_representation("real(kind=16)" REAL16_TYPE_Fortran LANGUAGE Fortran)
+
+  include(CheckFortranCompilerFlag)
+  check_fortran_compiler_flag("-fallow-argument-mismatch" ADIOS2_USE_Fortran_flag_argument_mismatch)
 endif()

